taskfiles/utils/cpp-lint.yaml

```
@@ -4,36 +4,324 @@ set: ["u", "pipefail"]
+            {{- end}}
+          \) \
+        {{- end}}
+        -print0
+      | xargs -0 clang-format \
+        {{- range .FLAGS}}
+          {{.}} \
+        {{- end}}
+        -Werror
 
-  # Runs clang-tidy on C++ files at the given paths.
+  # Run `clang-tidy` on `SOURCE_PATH` and write the output to `OUTPUT_PATH`. The output is appended
+  # with "Passed" or "Failed" depending on the result.
   #
-  # @param {string} FLAGS Any flags to pass to clang-tidy.
-  # @param {string[]} SRC_PATHS The paths on which to run clang-tidy.
-  # @param {string} VENV_DIR Python virtual environment where clang-tidy is installed.
+  # @param {string} OUTPUT_PATH Path to output log of `clang-tidy` for `SOURCE_PATH`.
+  # @param {string} SOURCE_PATH Source file path to run `clang-tidy` on.
+  # @param {string} VENV_DIR Python virtual environment where `clang-tidy` is installed.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
   clang-tidy:
     internal: true
+    label: "{{.TASK}}:{{.SOURCE_PATH}}"
+    sources: ["{{.SOURCE_PATH}}"]
+    vars:
+      FLAGS:
+        ref: "default (list) .FLAGS"
     requires:
-      vars: ["FLAGS", "SRC_PATHS", "VENV_DIR"]
-    cmd: |-
-      . "{{.VENV_DIR}}/bin/activate"
-      find {{- range .SRC_PATHS}} "{{.}}" {{- end}} \
-        -type f \
-        \( -iname "*.cpp" -o -iname "*.h" -o -iname "*.hpp" \) \
-        -print0 | \
-          xargs -0 clang-tidy {{.FLAGS}}
+      vars: ["OUTPUT_PATH", "SOURCE_PATH", "VENV_DIR"]
+    generates: ["{{.OUTPUT_PATH}}"]
+    cmds:
+      - defer: |-
+          if [[ 0 -eq "{{.EXIT_CODE}}" ]]; then
+            echo "Passed" >> "{{.OUTPUT_PATH}}"
+          else
+            echo "Failed" >> "{{.OUTPUT_PATH}}"
+          fi
+      - |-
+        . "{{.VENV_DIR}}/bin/activate"
+        clang-tidy \
+          {{- range .FLAGS}}
+            {{.}} \
+          {{- end}}
+          {{.SOURCE_PATH}} &> "{{.OUTPUT_PATH}}"
+
+  # Run `clang-tidy` in parallel on C++ source files that differ between HEAD and the target branch.
+  # On failure the `clang-tidy` output is printed. The task will fail on the first file to fail
+  # `clang-tidy`, which will result in some files having incomplete output. Either fix the known
+  # errors and run the task again or directly run `clang-tidy` on a file of interest.
+  #
+  # @param {string} VENV_DIR Python virtual environment where `clang-tidy` is installed.
+  # @param {string} [GIT_REF="origin/main"] A `git` reference that HEAD is diff'd with.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
+  # @param {string} [OUTPUT_DIR="{{.ROOT_DIR}}/lint-clang-tidy"] Directory to store the `clang-tidy`
+  # output for each file.
+  clang-tidy-diff:
+    internal: true
+    label: "{{.TASK}}:{{.OUTPUT_DIR}}"
+    vars:
+      GIT_REF: >-
+        {{default "origin/main" .GIT_REF}}
+      DIFF_SOURCE_PATHS_:
+        sh: >-
+          git diff --name-only --ignore-submodules "{{.GIT_REF}}"
+          | grep ".*\.\(cpp\|h\|hpp\)"
```

```suggestion
          git diff --name-only --ignore-submodules "{{.GIT_REF}}" **/*.{cpp,h,hpp}
```

That should work, right?

---

Pretty sure you're right, but I'll double check.